### PR TITLE
676 validate model properties error

### PR DESCRIPTION
**Pull request by CharlesRendle**
PR to satisfy ticket #676

Changes the information shown to the user in the event of a validation error by adding `property_path` and `offending_value` properties to `ValidateModelPropertiesError`.

`property_path` replaces the existing `property_name` property and takes the form of a list which drills down from the highest validated object to the offending property. 
Combining this with the `offending_value` should make it clear to the user where the error lies and what needs to be changed to satisfy the validation.

All relevant unit tests have been altered to reflect this change and some new ones added to demonstrate scenarios where property paths are constructed as properties belonging to higher level objects or when they are nested in lists.

Other changes:
- Adding truncation (to 50 chars) of offending values in the validation messages
- Updating references to the validation function so they all take the form `v.what_is_being_validated` E.g. `v.string`

### CODE REVIEW

**Review by robons — COMMENTED** [review]
Looks pretty close to me. Good work!

File: src/csvcubed/utils/validations.py [hunk 1]
```diff
@@ -8,164 +9,170 @@
...
+                property_path,
+                value,
             )
         ]
     elif isnan(value):
         return [
             ValidateModelPropertiesError(
-                "This variable should be a float value but is Not a Number (NaN), check the following variable:",
-                property_name,
+                f"The value '{truncate(str(value), 50)}' should be a float but is Not a Number (NaN). Check the following variable at the property path: '{property_path}'",
```
[the remark this answers is not in the document]
Comment (CharlesRendle) [reply src/csvcubed/utils/validations.py]: Removed the truncated value from the message.

File: src/csvcubed/utils/validations.py [hunk 2]
```diff
@@ -174,17 +181,19 @@ def validate_file(
...
                 )
             ]
         else:
             return []
     else:
         return [
             ValidateModelPropertiesError(
-                "This is not a valid file path, check the following variable:",
-                property_name,
+                f"The file '{truncate(str(value), 50)}' is not a valid file path. Check the following variable at the property path: '{property_path}'",
```
[the remark this answers is not in the document]
Comment (CharlesRendle) [reply src/csvcubed/utils/validations.py]: Added this portion of the message to the `__post_init__` of the `ValidateModelPropertiesError` class

[the remark this answers is not in the document]
Comment (CharlesRendle) [reply src/csvcubed/utils/validations.py]: and removed as repeated strings in each of the validation functions.

File: tests/unit/utils/test_validations.py [hunk 3]
```diff
@@ -767,6 +828,110 @@ def test_boolean_incorrect():
...
+    assert "This value '5' should be a boolean value." in errors[0].message
+    assert errors[0].property_path == ["bool_test_variable"]
+    assert errors[0].offending_value == 5
+
+
+def test_primative_inside_list_inside_object():
+    """
+    Test to ensure that when a primative inside a list within a nested object is
+    incorrect the returned property_path contains elements corresponding to the
+    depth property in the nested object.
```
[the remark this answers is not in the document]
Comment (CharlesRendle) [reply tests/unit/utils/test_validations.py]: Changed the description to `... corresponding to the
    offending property's depth in the ...`

File: src/csvcubed/models/validationerror.py [hunk 4]
```diff
@@ -69,4 +69,11 @@ class ValidateModelPropertiesError(ValidationError):
     This error will be returned with the name of the variable that has not been validated.
     """
 
-    property_name: str
+    property_path: List[str]
+    offending_value: Any
+
+    def __post_init__(self):
+        self.message = (
```
Comment (robons) [inline src/csvcubed/models/validationerror.py:76]: you should use `self.message += ...` to make it a bit terser.

File: src/csvcubed/utils/validations.py [hunk 5]
```diff
@@ -8,164 +9,170 @@
...
+                property_path,
+                value,
             )
         ]
     elif isinf(value):
         return [
             ValidateModelPropertiesError(
-                "This variable should be a float value but is +-infinity, check the following variable:",
-                property_name,
+                "The value should be a float but is +-infinity.",
```
Comment (robons) [inline src/csvcubed/utils/validations.py:166]: `±`

File: src/csvcubed/models/validationerror.py (hunk 4 above)
Comment (CharlesRendle) [reply src/csvcubed/models/validationerror.py]: Changed

File: src/csvcubed/utils/validations.py (hunk 5 above)
Comment (CharlesRendle) [reply src/csvcubed/utils/validations.py]: Changed

**Pull request merged**